crates/bevy_ecs/src/world/archetype_invariants.rs

```diff
@@ -0,0 +1,618 @@
+        }
+    }
+
+    /// Creates an archetype invariant where components of `B` can only appear with each other.
+    ///
+    /// In other words, if any component of this bundle is present, then _only_ components from this bundle can be present.
+    #[inline]
+    pub fn exclusive() -> Self {
+        Self {
+            premise: ArchetypeStatement::<B>::any_of(),
+            consequence: ArchetypeStatement::<B>::only(),
+        }
+    }
+}
+
+/// A statement about the presence or absence of some subset of components in the given [`Bundle`]
+///
+/// This type is used as part of an [`ArchetypeInvariant`].
+///
+/// When used as a premise, the archetype invariant matches all entities which satisfy the statement.
+/// When used as a consequence, then the statment must be true for all entities that were matched by the premise.
+///
+/// For the statements about a single component `C`, wrap it in a single-component bundle `(C,)`.
+/// For single component bundles, `AllOf` and `AnyOf` are equivalent.
+/// Prefer `ArchetypeStatement::<(C,)>::all_of` over `ArchetypeStatement::<(C,)>::any_of` for consistency and clarity.
+///
+/// Note that this is converted to an [`UntypedArchetypeStatement`] when added to a [`World`].
+/// This is to ensure compatibility between different invariants.
+#[derive(Clone, Debug, PartialEq)]
+pub enum ArchetypeStatement<B: Bundle> {
+    /// Evaluates to true if and only if the entity has all of the components present in the bundle `B`.
+    AllOf(PhantomData<B>),
+    /// The entity has at least one component in the bundle `B`.
+    /// When using a single-component bundle, `AllOf` is preferred.
+    AnyOf(PhantomData<B>),
+    /// The entity has zero or one of the components in the bundle `B`, but no more.
+    /// When using a single-component bundle, this will always be true.
+    AtMostOneOf(PhantomData<B>),
+    /// The entity has none of the components in the bundle `B`.
+    NoneOf(PhantomData<B>),
+    /// The entity contains only components from the bundle `B`, and no others.
+    Only(PhantomData<B>),
+}
+
+impl<B: Bundle> ArchetypeStatement<B> {
+    /// Erases the type information of this archetype statement.
+    ///
+    /// Requires mutable world access, since the components might not have been added to the world yet.
+    pub fn into_untyped(self, world: &mut World) -> UntypedArchetypeStatement {
+        let component_ids = B::component_ids(&mut world.components, &mut world.storages);
+        let component_ids: HashSet<ComponentId> = component_ids.into_iter().collect();
+
+        match self {
+            ArchetypeStatement::AllOf(_) => UntypedArchetypeStatement::AllOf(component_ids),
+            ArchetypeStatement::AnyOf(_) => {
+                if component_ids.len() == 1 {
+                    warn!("An `ArchetypeStatement::AnyOf` was constructed for a bundle with only one component. Prefer the equivalent `ArchetypeStatment:AllOf` for consistency and clarity.");
```

Comment:
I think that on the balance this isn't worth warning about.

I'm particularly nervous about warnings in cases where these things are automatically generated.

Reply:
I agree, it also probably introduces an unnecessary branching instruction in release mode.

Reply:
Ah, I hadn't considered the auto-generated statement. Totally agreed- this should be just removed. Fixed as of 22b6af1e36a4ea5a54982c3ddc949212cd18eb2f